ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ExternalTableCopyTaskBuilder.java

```diff
@@ -54,14 +59,53 @@
   List<Task<? extends Serializable>> tasks(TaskTracker tracker) {
     List<Task<? extends Serializable>> tasks = new ArrayList<>();
     Iterator<DirCopyWork> itr = work.getPathsToCopyIterator();
-    while (tracker.canAddMoreTasks() && itr.hasNext()) {
+    int numTaskCanBeAdded = tracker.numTaskCanBeAdded();
+    Task<? extends Serializable> barrierTask = TaskFactory.get(new DependencyCollectionWork(), conf);
+    while (numTaskCanBeAdded-- > 0 && itr.hasNext()) {
       DirCopyWork dirCopyWork = itr.next();
       Task<DirCopyWork> task = TaskFactory.get(dirCopyWork, conf);
       tasks.add(task);
-      tracker.addTask(task);
+      barrierTask.addDependentTask(task);
       LOG.debug("added task for {}", dirCopyWork);
     }
-    return tasks;
+
+    if (!tasks.isEmpty()) {
+      tracker.addDependentTask(barrierTask);
+      tracker.addTaskList(tasks);
+      return Collections.singletonList(barrierTask);
+    } else {
+      return tasks;
+    }
+  }
+
+  private static Integer setTargetPathOwnerInt(Path targetPath, Path sourcePath, HiveConf conf) throws IOException {
+    FileSystem targetFs = targetPath.getFileSystem(conf);
+    if (!targetFs.exists(targetPath)) {
+      targetFs.create(targetPath);
+    }
+    FileStatus status = sourcePath.getFileSystem(conf).getFileStatus(sourcePath);
+    if (status == null) {
+      throw new IOException("source path missing " + sourcePath);
+    }
+    targetPath.getFileSystem(conf).setOwner(targetPath, status.getOwner(), status.getGroup());
+    return null;
+  }
+
+  private static Integer setTargetPathOwner(Path targetPath, Path sourcePath, HiveConf conf, String distCpDoAsUser)
+          throws IOException {
+    if (distCpDoAsUser == null) {
```

Comment:
i dont think distCpDoAsUser can be null since we will always need a user to do distcp which would be the beacon super user.

Reply:
same as distcp below ..i think better to handle it